File: native/macos/Sources/MightyCore/ToolkitStore.swift
import CryptoKit
import Foundation

// MARK: - Approval

/// Approval bound to the entry's canonical content hash.  For repoScript the
/// resolved commit SHA (always 40-hex) is stored alongside the content hash.
public struct ToolkitApproval: Sendable, Equatable {
    public let contentHash: String
    public let resolvedCommit: String?

    public init(contentHash: String, resolvedCommit: String? = nil) {
        self.contentHash = contentHash
        self.resolvedCommit = resolvedCommit
    }
}

// MARK: - Executor

/// Injected into approve() so tests never hit the network or the real CLIs.
public protocol ToolkitCommandExecutor: Sendable {
    func run(_ argv: [String]) throws -> String
}

// MARK: - Error

public struct ToolkitStoreError: Error, Sendable {
    public let message: String
    public init(_ message: String) { self.message = message }
}

// MARK: - ToolkitStore

/// Manages the user's toolkit list next to workspace-state.json.
///
/// `list()` always returns bundled entries; if toolkit.json exists but is
/// unreadable or structurally invalid the error is surfaced through the second
/// return value and the file is never modified.
public actor ToolkitStore {
    private let fileURL: URL

    // Internal state set once on first access.
    private var loaded = false
    private var userEntries: [ToolkitEntry] = []
    private var approvals: [String: ToolkitApproval] = [:]
    private var fileError: (any Error)?
    // Raw entry dicts (including unknown keys like a stale `platforms`) for
    // entries that were loaded from disk and have not been mutated this session.
    private var rawEntryDicts: [String: [String: Any]] = [:]
    // IDs of entries added, removed, edited, approved or imported this session.
    // Those entries are written in canonical form; untouched entries use rawEntryDicts.
    private var touchedEntryIds: Set<String> = []

    public init(directory: URL) {
        self.fileURL = directory.appendingPathComponent("toolkit.json")
    }

    // MARK: - List

    /// Bundled entries are always first.  If toolkit.json is invalid, user
    /// entries are empty and the error is non-nil; the file is left untouched.
    /// Other-OS entries (e.g. winget on macOS) are kept in storage but omitted
    /// from the returned list.
    public func list() -> (entries: [ToolkitEntry], error: (any Error)?) {
        doLoad()
        let visible = userEntries.filter { $0.install.platforms.contains(.macOS) }
        return (ToolkitBundled.entries + visible, fileError)
    }

    // MARK: - Approval

    /// Returns the stored approval if and only if it still matches the entry's
    /// current canonical hash.  Returns nil for bundled entries and any entry
    /// whose content changed since approval.
    public func approval(for entry: ToolkitEntry) -> ToolkitApproval? {
        doLoad()
        guard entry.source == .user, let stored = approvals[entry.entryId] else { return nil }
        return stored.contentHash == Self.canonicalHash(entry) ? stored : nil
    }

    /// Approves a user entry.  For a repoScript with a tag ref the executor
    /// resolves it to a full 40-hex commit SHA via `git ls-remote`.
    public func approve(entryId: String, executor: any ToolkitCommandExecutor) throws {
        try requireLoaded()
        guard let entry = userEntries.first(where: { $0.entryId == entryId }) else {
            throw ToolkitStoreError("Entry not found: \(entryId)")
        }
        let hash = Self.canonicalHash(entry)
        var resolvedCommit: String? = nil
        if case .repoScript(let url, let ref, _) = entry.install {
            if Self.is40HexSHA(ref) {
                resolvedCommit = ref
            } else {
                let output = try executor.run(["git", "ls-remote", url, ref])
                resolvedCommit = try Self.parseSHA(from: output, tag: ref)
            }
        }
        approvals[entryId] = ToolkitApproval(contentHash: hash, resolvedCommit: resolvedCommit)
        rawEntryDicts.removeValue(forKey: entryId)
        touchedEntryIds.insert(entryId)
        try persist()
    }

    // MARK: - Mutate

    /// Adds or replaces a user entry.  The new entry is always unapproved.
    public func addEntry(_ entry: ToolkitEntry) throws {
        try requireLoaded()
        userEntries.removeAll { $0.entryId == entry.entryId }
        approvals.removeValue(forKey: entry.entryId)
        rawEntryDicts.removeValue(forKey: entry.entryId)
        touchedEntryIds.insert(entry.entryId)
        var e = entry; e.source = .user
        userEntries.append(e)
        try persist()
    }

    /// Removes a user entry.  No command is executed and no file other than
    /// toolkit.json is touched.
    public func removeEntry(id: String) throws {
        try requireLoaded()
        userEntries.removeAll { $0.entryId == id }
        approvals.removeValue(forKey: id)
        rawEntryDicts.removeValue(forKey: id)
        touchedEntryIds.insert(id)
        try persist()
    }

    // MARK: - Export / Import

    /// Serialises user entries as a JSON array with no approval data.
    public func exportData() throws -> Data {
        try requireLoaded()
        let objects = userEntries.map { Self.entryToObject($0) }
        return try JSONSerialization.data(withJSONObject: objects, options: [.sortedKeys])
    }

    /// Adds every entry from a JSON array as unapproved.  An existing id is
    /// replaced and becomes unapproved.
    public func importData(_ data: Data) throws {
        try requireLoaded()
        guard let array = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]] else {
            throw ToolkitStoreError("Import data must be a JSON array")
        }
        for rawObject in array {
            var stripped = rawObject
            stripped.removeValue(forKey: "approval")
            let entry = try ToolkitEntryDecoder.decode(stripped)
            userEntries.removeAll { $0.entryId == entry.entryId }
            approvals.removeValue(forKey: entry.entryId)
            rawEntryDicts.removeValue(forKey: entry.entryId)
            touchedEntryIds.insert(entry.entryId)
            userEntries.append(entry)
        }
        try persist()
    }

    // MARK: - Internal

    private func doLoad() {
        guard !loaded else { return }
        loaded = true
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return }
        guard let data = try? Data(contentsOf: fileURL),
              let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              object["version"] as? Int == 1 else {
            fileError = ToolkitStoreError("toolkit.json을 읽을 수 없습니다. 원본 파일은 변경하지 않았습니다.")
            return
        }
        for rawEntry in (object["entries"] as? [[String: Any]]) ?? [] {
            // Store the complete raw dict (including unknown keys like a stale
            // `platforms` and the existing approval) before any stripping.
            if let entryId = rawEntry["id"] as? String, !entryId.isEmpty {
                rawEntryDicts[entryId] = rawEntry
            }
            var raw = rawEntry
            let rawApproval = raw.removeValue(forKey: "approval") as? [String: Any]
            guard let entry = try? ToolkitEntryDecoder.decode(raw) else { continue }
            userEntries.append(entry)
            if let a = rawApproval, let hash = a["contentHash"] as? String {
                approvals[entry.entryId] = ToolkitApproval(
                    contentHash: hash, resolvedCommit: a["resolvedCommit"] as? String)
            }
        }
    }

    private func requireLoaded() throws {
        doLoad()
        if let err = fileError { throw err }
    }

    private func persist() throws {
        var objects: [[String: Any]] = []
        for entry in userEntries {
            if !touchedEntryIds.contains(entry.entryId),
               let rawDict = rawEntryDicts[entry.entryId] {
                // Untouched entry: emit original dict verbatim (preserves unknown
                // keys such as a stale `platforms` and the stored approval).
                objects.append(rawDict)
            } else {
                // Touched entry: emit canonical form with current approval.
                var obj = Self.entryToObject(entry)
                if let approval = approvals[entry.entryId] {
                    var a: [String: Any] = ["contentHash": approval.contentHash]
                    if let sha = approval.resolvedCommit { a["resolvedCommit"] = sha }
                    obj["approval"] = a
                }
                objects.append(obj)
            }
        }
        let fileObject: [String: Any] = ["version": 1, "entries": objects]
        // Canonical file form: sorted keys, 2-space indentation, LF, trailing newline.
        var payload = try JSONSerialization.data(
            withJSONObject: fileObject, options: [.sortedKeys, .prettyPrinted])
        if payload.last != UInt8(ascii: "\n") { payload.append(UInt8(ascii: "\n")) }
        let dir = fileURL.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let tmp = dir.appendingPathComponent("toolkit-" + UUID().uuidString + ".tmp")
        defer { try? FileManager.default.removeItem(at: tmp) }
        try payload.write(to: tmp, options: .atomic)
        _ = try FileManager.default.replaceItemAt(fileURL, withItemAt: tmp)
    }

    // MARK: - Static helpers (nonisolated)

    nonisolated static func canonicalHash(_ entry: ToolkitEntry) -> String {
        let obj = entryToObject(entry)
        guard let data = try? JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys]) else { return "" }
        return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    /// The exact UTF-8 bytes that are fed into the SHA-256 hash: sorted keys,
    /// no insignificant whitespace, approval and unknown keys excluded.
    /// Matches BuildCanonicalJson on Windows.
    nonisolated public static func canonicalJson(_ entry: ToolkitEntry) -> String {
        let obj = entryToObject(entry)
        guard let data = try? JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys]),
              let json = String(data: data, encoding: .utf8) else { return "" }
        return json
    }

    nonisolated static func entryToObject(_ entry: ToolkitEntry) -> [String: Any] {
        ["id": entry.entryId, "displayName": entry.displayName,
         "install": installToObject(entry.install)]
    }

    private nonisolated static func installToObject(_ spec: ToolkitInstallSpec) -> [String: Any] {
        switch spec {
        case .plugin(let source, let pluginID):
            return ["kind": "plugin", "source": source, "pluginID": pluginID]
        case .mcp(let name, let executable, let args):
            return ["kind": "mcp", "name": name, "executable": executable, "args": args]
        case .skill(let url):
            return ["kind": "skill", "url": url]
        case .package(let manager, let name, let executable):
            var obj: [String: Any] = ["kind": "package", "manager": manager.rawValue, "name": name]
            if let exe = executable { obj["executable"] = exe }
            return obj
        case .repoScript(let url, let ref, let scriptPath):
            return ["kind": "repoScript", "url": url, "ref": ref, "scriptPath": scriptPath]
        }
    }

    nonisolated static func is40HexSHA(_ s: String) -> Bool {
        s.count == 40 && s.unicodeScalars.allSatisfy {
            ($0.value >= 0x30 && $0.value <= 0x39) ||
            ($0.value >= 0x61 && $0.value <= 0x66)
        }
    }

    private nonisolated static func parseSHA(from output: String, tag: String) throws -> String {
        for line in output.split(separator: "\n") {
            let sha = String(line.split(separator: "\t").first ?? Substring()).trimmingCharacters(in: .whitespaces)
            if is40HexSHA(sha) { return sha }
        }
        throw ToolkitStoreError("tag '\(tag)'를 커밋 SHA로 해석할 수 없습니다.")
    }
}

File: native/windows/MightyClaude.Core/ToolkitStore.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using ToolkitFileEntry = MightyClaude.Core.ToolkitFileReader.ToolkitFileEntry;

namespace MightyClaude.Core;

/// <summary>
/// Manages the user's toolkit list next to workspace-state.json.
///
/// list() always returns bundled entries first, then this-OS user entries.
/// Other-OS entries (brew/repoScript on Windows) are preserved in the file
/// byte-identical and never shown in the list.
/// If toolkit.json exists but is unreadable the error is surfaced and the file
/// is never modified.
/// </summary>
public sealed class ToolkitStore
{
    private readonly string directory;

    private bool loaded;
    private List<ToolkitFileEntry> thisOsEntries = [];
    private Dictionary<string, ToolkitFileReader.ToolkitApproval> approvals = [];
    private string? fileError;
    // Ordered list of all entry IDs as read from the file (both platforms).
    private List<string> allEntryIds = [];
    // Raw JSON text for each entry, keyed by id (preserves unknown keys like `platforms`).
    private Dictionary<string, string> allEntryRawText = [];
    // IDs of entries that are other-OS (brew/repoScript on Windows).
    private HashSet<string> otherOsIds = [];
    // IDs of entries added, edited, approved or removed this session.
    // Touched entries are written canonically; untouched entries emit original bytes.
    private HashSet<string> touchedEntryIds = [];

    public ToolkitStore(string directory) => this.directory = directory;

    private string FilePath => Path.Combine(directory, "toolkit.json");

    // Bundled entry — always first, always visible, never requires approval.
    public static IReadOnlyList<ToolkitFileEntry> Bundled { get; } = [
        new("mighty-styles", "mighty-styles", "plugin",
            new ToolkitFileReader.PluginSpec("athens96/mighty-styles", "mighty-styles@mighty-styles"),
            Approval: null,
            Source: ToolkitFileReader.ToolkitEntrySource.Bundled),
    ];

    // ── Public API ───────────────────────────────────────────────────────────

    /// Returns (bundled + this-OS user entries, error). Other-OS entries are
    /// kept in the file but never appear in the list.
    public (IReadOnlyList<ToolkitFileEntry> Entries, string? Error) List()
    {
        DoLoad();
        return ([..Bundled, ..thisOsEntries], fileError);
    }

    /// Returns the stored approval for a user entry only if the content hash
    /// still matches. Returns null for bundled entries or if the entry changed.
    public ToolkitFileReader.ToolkitApproval? GetApproval(ToolkitFileEntry entry)
    {
        DoLoad();
        if (entry.Source == ToolkitFileReader.ToolkitEntrySource.Bundled) return null;
        if (!approvals.TryGetValue(entry.Id, out var stored)) return null;
        return stored.ContentHash == CanonicalHash(entry) ? stored : null;
    }

    /// Approves a user entry by computing its current canonical hash.
    /// resolvedCommit is required for repoScript entries (40-hex SHA).
    public void Approve(string id, string? resolvedCommit = null)
    {
        RequireLoaded();
        var entry = thisOsEntries.FirstOrDefault(e => e.Id == id)
            ?? throw new InvalidOperationException("Entry not found: " + id);
        approvals[id] = new ToolkitFileReader.ToolkitApproval(CanonicalHash(entry), resolvedCommit);
        touchedEntryIds.Add(id);
        Persist();
    }

    /// Adds or replaces a user entry (always unapproved after this call).
    public void Add(ToolkitFileEntry entry)
    {
        RequireLoaded();
        thisOsEntries.RemoveAll(e => e.Id == entry.Id);
        approvals.Remove(entry.Id);
        touchedEntryIds.Add(entry.Id);
        thisOsEntries.Add(entry with { Approval = null, Source = ToolkitFileReader.ToolkitEntrySource.User });
        Persist();
    }

    /// Serialises user entries (without approvals) to a toolkit.json string for export.
    public string Export()
    {
        DoLoad();
        var parts = thisOsEntries.Select(EntryToJson).ToList();
        return $"{{\"version\":1,\"entries\":[{string.Join(",", parts)}]}}";
    }

    /// Removes a user entry. No command is executed.
    public void Remove(string id)
    {
        RequireLoaded();
        thisOsEntries.RemoveAll(e => e.Id == id);
        approvals.Remove(id);
        touchedEntryIds.Add(id);
        Persist();
    }

    // ── Load / persist ────────────────────────────────────────────────────────

    private void DoLoad()
    {
        if (loaded) return;
        loaded = true;
        var filePath = FilePath;
        if (!File.Exists(filePath)) return;
        try
        {
            var json = File.ReadAllText(filePath);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) { fileError = Locale.Get("settings.toolkit.errorBanner"); return; }
            if (!root.TryGetProperty("version", out var ver) || !ver.TryGetInt32(out int version) || version != 1) { fileError = Locale.Get("settings.toolkit.errorBanner"); return; }
            if (!root.TryGetProperty("entries", out var entriesEl) || entriesEl.ValueKind != JsonValueKind.Array) { fileError = Locale.Get("settings.toolkit.errorBanner"); return; }

            foreach (var item in entriesEl.EnumerateArray())
            {
                if (item.ValueKind != JsonValueKind.Object) continue;
                if (!item.TryGetProperty("id", out var idProp) || idProp.ValueKind != JsonValueKind.String) continue;
                var itemId = idProp.GetString() ?? "";
                if (itemId.Length == 0) continue;
                allEntryIds.Add(itemId);
                allEntryRawText[itemId] = item.GetRawText();
                if (IsOtherOsItem(item))
                {
                    otherOsIds.Add(itemId);
                    continue;
                }
                // Strip the approval field before decoding (store manages it separately).
                var entry = ToolkitFileReader.ParseEntry(item);
                if (entry is null) continue;
                thisOsEntries.Add(entry with { Approval = null, Source = ToolkitFileReader.ToolkitEntrySource.User });
                if (item.TryGetProperty("approval", out var approvalEl) && approvalEl.ValueKind == JsonValueKind.Object)
                {
                    if (approvalEl.TryGetProperty("contentHash", out var hashEl) && hashEl.ValueKind == JsonValueKind.String)
                    {
                        var hash = hashEl.GetString() ?? "";
                        string? commit = null;
                        if (approvalEl.TryGetProperty("resolvedCommit", out var commitEl) && commitEl.ValueKind == JsonValueKind.String)
                            commit = commitEl.GetString();
                        if (hash.Length > 0) approvals[entry.Id] = new ToolkitFileReader.ToolkitApproval(hash, commit);
                    }
                }
            }
        }
        catch { fileError = Locale.Get("settings.toolkit.errorBanner"); }
    }

    private void RequireLoaded()
    {
        DoLoad();
        if (fileError is not null) throw new InvalidOperationException(fileError);
    }

    private void Persist()
    {
        var dir = directory;
        Directory.CreateDirectory(dir);
        var thisOsIds = new HashSet<string>(thisOsEntries.Select(e => e.Id));
        var entryLines = new List<string>();

        // Emit entries in file order; untouched entries use original bytes.
        foreach (var id in allEntryIds)
        {
            if (otherOsIds.Contains(id))
            {
                if (allEntryRawText.TryGetValue(id, out var rawText)) entryLines.Add(rawText);
            }
            else if (touchedEntryIds.Contains(id))
            {
                if (thisOsIds.Contains(id))
                    entryLines.Add(BuildEntryLine(thisOsEntries.First(e => e.Id == id)));
                // else: removed — skip
            }
            else
            {
                // Untouched thisOS entry: emit original bytes if still present.
                if (thisOsIds.Contains(id) && allEntryRawText.TryGetValue(id, out var rawText))
                    entryLines.Add(rawText);
            }
        }

        // New entries added this session (not in the original file).
        foreach (var entry in thisOsEntries)
            if (!allEntryIds.Contains(entry.Id))
                entryLines.Add(BuildEntryLine(entry));

        // Canonical file form: sorted top-level keys, 2-space indent, LF, one trailing newline.
        var body = entryLines.Count > 0
            ? "\n    " + string.Join(",\n    ", entryLines) + "\n  "
            : "";
        var json = "{\n  \"entries\": [" + body + "],\n  \"version\": 1\n}\n";
        var tmpPath = FilePath + "." + Path.GetRandomFileName() + ".tmp";
        try
        {
            File.WriteAllText(tmpPath, json, Encoding.UTF8);
            File.Move(tmpPath, FilePath, overwrite: true);
        }
        finally { if (File.Exists(tmpPath)) try { File.Delete(tmpPath); } catch { } }
    }

    // Builds one entry line for Persist(): canonical form with sorted keys, approval first.
    private string BuildEntryLine(ToolkitFileEntry entry)
    {
        var install = BuildInstallJson(entry);
        if (approvals.TryGetValue(entry.Id, out var approval))
        {
            var a = approval.ResolvedCommit is null
                ? $"\"contentHash\":{Jstr(approval.ContentHash)}"
                : $"\"contentHash\":{Jstr(approval.ContentHash)},\"resolvedCommit\":{Jstr(approval.ResolvedCommit)}";
            // Keys sorted: approval < displayName < id < install
            return $"{{\"approval\":{{{a}}},\"displayName\":{Jstr(entry.DisplayName)},\"id\":{Jstr(entry.Id)},\"install\":{install}}}";
        }
        return $"{{\"displayName\":{Jstr(entry.DisplayName)},\"id\":{Jstr(entry.Id)},\"install\":{install}}}";
    }

    // ── Canonical hash ────────────────────────────────────────────────────────

    /// SHA-256 of the canonical JSON for the entry (no approval, sorted keys).
    /// Matches ToolkitStore.canonicalHash(_:) on macOS.
    public static string CanonicalHash(ToolkitFileEntry entry)
    {
        var json = BuildCanonicalJson(entry);
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(json));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    private static string BuildCanonicalJson(ToolkitFileEntry entry)
    {
        // Alphabetical key order: displayName, id, install
        return $"{{\"displayName\":{Jstr(entry.DisplayName)},\"id\":{Jstr(entry.Id)},\"install\":{BuildInstallJson(entry)}}}";
    }

    private static string BuildInstallJson(ToolkitFileEntry entry) => entry.Install switch
    {
        // Sorted keys: kind, pluginID, source
        ToolkitFileReader.PluginSpec ps =>
            $"{{\"kind\":\"plugin\",\"pluginID\":{Jstr(ps.PluginId)},\"source\":{Jstr(ps.Source)}}}",
        // Sorted keys: args, executable, kind, name
        ToolkitFileReader.McpSpec ms =>
            $"{{\"args\":{JsonSerializer.Serialize(ms.Args, Wire.Json)},\"executable\":{Jstr(ms.Executable)},\"kind\":\"mcp\",\"name\":{Jstr(ms.Name)}}}",
        // Sorted keys: kind, url
        ToolkitFileReader.SkillSpec ss =>
            $"{{\"kind\":\"skill\",\"url\":{Jstr(ss.Url)}}}",
        // Sorted keys: executable, kind, manager, name  (winget has executable)
        ToolkitFileReader.PackageSpec { Manager: "winget", Executable: { } exe } pks =>
            $"{{\"executable\":{Jstr(exe)},\"kind\":\"package\",\"manager\":\"winget\",\"name\":{Jstr(pks.Name)}}}",
        // Sorted keys: kind, manager, name  (brew/npm)
        ToolkitFileReader.PackageSpec pks =>
            $"{{\"kind\":\"package\",\"manager\":{Jstr(pks.Manager)},\"name\":{Jstr(pks.Name)}}}",
        // Sorted keys: kind, ref, scriptPath, url
        ToolkitFileReader.RepoScriptSpec rs =>
            $"{{\"kind\":\"repoScript\",\"ref\":{Jstr(rs.Ref)},\"scriptPath\":{Jstr(rs.ScriptPath)},\"url\":{Jstr(rs.Url)}}}",
        _ => "{}",
    };

    // Serialize a string to a JSON string literal (with proper escaping).
    private static string Jstr(string? s) => JsonSerializer.Serialize(s ?? "", Wire.Json);

    // ── JSON serialization of an entry (no approval) ─────────────────────────

    private static string EntryToJson(ToolkitFileEntry entry)
    {
        // Alphabetical key order: displayName, id, install (approval added later by Persist)
        return $"{{\"displayName\":{Jstr(entry.DisplayName)},\"id\":{Jstr(entry.Id)},\"install\":{BuildInstallJson(entry)}}}";
    }

    // ── Platform helpers ─────────────────────────────────────────────────────

    // brew and repoScript are macOS-only; everything else (plugin, mcp, skill, npm, winget) runs on Windows.
    private static bool IsOtherOsItem(JsonElement item)
    {
        if (!item.TryGetProperty("install", out var inst) || inst.ValueKind != JsonValueKind.Object) return false;
        if (!inst.TryGetProperty("kind", out var kindEl) || kindEl.ValueKind != JsonValueKind.String) return false;
        var kind = kindEl.GetString() ?? "";
        if (kind == "repoScript") return true;
        if (kind == "package" && inst.TryGetProperty("manager", out var mgrEl) && mgrEl.ValueKind == JsonValueKind.String)
            return mgrEl.GetString() == "brew";
        return false;
    }
}
